File: app/src/main/java/hm/orz/chaos114/android/tumekyouen/model/Line.java
package hm.orz.chaos114.android.tumekyouen.model;

import com.google.auto.value.AutoValue;

/**
 * Ax+By+C=0を表現するクラス。
 */
@AutoValue
public abstract class Line {
    public abstract Point p1();

    public abstract Point p2();

    public abstract double a();

    public abstract double b();

    public abstract double c();

    public static Line create(Point p1, Point p2) {
        double a = p1.y() - p2.y();
        double b = p2.x() - p1.x();
        double c = p1.x() * p2.y() - p2.x() * p1.y();

        return new AutoValue_Line(p1, p2, a, b, c);
    }

    public double getY(double x) {
        return -1 * (a() * x + c()) / b();
    }

    public double getX(double y) {
        return -1 * (b() * y + c()) / a();
    }
}

File: app/src/main/java/hm/orz/chaos114/android/tumekyouen/model/Line.kt
package hm.orz.chaos114.android.tumekyouen.model

/**
 * a*x + b*y + c = 0
 */
data class Line(
        val p1: Point,
        val p2: Point,
        val a: Double,
        val b: Double,
        val c: Double
) {
    fun getY(x: Double): Double {
        return -1 * (a * x + c) / b
    }

    fun getX(y: Double): Double {
        return -1 * (b * y + c) / a
    }

    companion object {

        fun create(p1: Point, p2: Point): Line {
            val a = p1.y - p2.y
            val b = p2.x - p1.x
            val c = p1.x * p2.y - p2.x * p1.y

            return Line(p1, p2, a, b, c)
        }
    }
}
